router/batchrouter/asyncdestinationmanager/marketo-bulk-upload/marketobulkupload.go

```diff
@@ -0,0 +1,325 @@
+	metadata, ok := failedJobsResponse["metadata"].(map[string]interface{})
+	if !ok {
+		// TODO
+		pkgLogger.Errorf("[Batch Router] Failed to typecast failed jobs response for Dest Type %v with statusCode %v and body %v", "MARKETO_BULK_UPLOAD", internalStatusCode, string(failedBodyBytes))
+		return common.GetUploadStatsResponse{
+			Status: "500",
+		}
+	}
+
+	failedKeys, errFailed := misc.ConvertStringInterfaceToIntArray(metadata["failedKeys"])
+	warningKeys, errWarning := misc.ConvertStringInterfaceToIntArray(metadata["warningKeys"])
+	succeededKeys, errSuccess := misc.ConvertStringInterfaceToIntArray(metadata["succeededKeys"])
+
+	// Build the response body
+	return common.GetUploadStatsResponse{
+		Status: internalStatusCode,
+		Metadata: common.EventStatMeta{
+			FailedKeys:    failedKeys,
+			ErrFailed:     errFailed,
+			WarningKeys:   warningKeys,
+			ErrWarning:    errWarning,
+			SucceededKeys: succeededKeys,
+			ErrSuccess:    errSuccess,
+		},
+	}
+}
+
+func ExtractJobStats(keyMap map[string]interface{}, importingJobIDs []int64) ([]int64, []int64) {
+	if keyMap == nil {
+		return []int64{}, importingJobIDs
+	}
+
+	_, ok := keyMap["successfulJobs"].([]interface{})
+	var succesfulJobIDs, failedJobIDsTrans []int64
+	var err error
+	if ok {
+		succesfulJobIDs, err = misc.ConvertStringInterfaceToIntArray(keyMap["successfulJobs"])
+		if err != nil {
+			failedJobIDsTrans = importingJobIDs
+		}
+	}
+	_, ok = keyMap["unsuccessfulJobs"].([]interface{})
+	if ok {
+		failedJobIDsTrans, err = misc.ConvertStringInterfaceToIntArray(keyMap["unsuccessfulJobs"])
+		if err != nil {
+			failedJobIDsTrans = importingJobIDs
+		}
+	}
+	return succesfulJobIDs, failedJobIDsTrans
+}
+
+func (b *MarketoBulkUploader) Upload(destination *backendconfig.DestinationT, asyncDestStruct *common.AsyncDestinationStruct) common.AsyncUploadOutput {
+	resolveURL := func(base, relative string) string {
+		// var logger logger.Logger
+		baseURL, _ := url.Parse(base)
+		// if err != nil {
+		// 	logger.Fatal(err)
+		// }
+		relURL, _ := url.Parse(relative)
+		// if err != nil {
+		// 	logger.Fatal(err)
+		// }
+		destURL := baseURL.ResolveReference(relURL).String()
+		return destURL
+	}
+	destinationID := destination.ID
+	destinationUploadUrl := asyncDestStruct.URL
+	url := resolveURL(b.TransformUrl, destinationUploadUrl)
+	filePath := asyncDestStruct.FileName
+	destConfig := destination.Config
+	destType := destination.DestinationDefinition.Name
+	failedJobIDs := asyncDestStruct.FailedJobIDs
+	importingJobIDs := asyncDestStruct.ImportingJobIDs
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		panic("BRT: Read File Failed" + err.Error())
+	}
+	defer file.Close()
+	var input []common.AsyncJob
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var tempJob common.AsyncJob
+		jobBytes := scanner.Bytes()
+		err := json.Unmarshal(jobBytes, &tempJob)
+		if err != nil {
+			panic("Unmarshalling a Single Line Failed")
```

Comment:
Can we fail the job instead of panicking?

Reply:
done